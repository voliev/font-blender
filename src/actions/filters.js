import {
  SET_CATEGORIES_FILTER,
  SET_SORTING_ORDER,
  SET_SORTING_PARAM,
  SET_TITLE_FILTER
} from "../constants";

export const setTitleFilter = (title = "") => ({
  type: SET_TITLE_FILTER,
  title
});

export const setCategoriesFilter = category => ({
  type: SET_CATEGORIES_FILTER,
  category
});

export const setSortingParam = param => ({
  type: SET_SORTING_PARAM,
  param
});

export const setSortingOrder = order => ({
  type: SET_SORTING_ORDER,
  order
});
