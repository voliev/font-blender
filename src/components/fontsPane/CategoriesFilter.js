import React, { Component } from "react";
import classNames from "classnames";
import FilterList from "@material-ui/icons/FilterList";
import OutsideClickHandler from "react-outside-click-handler";
import CategoriesList from "./CategoriesList";
import "./styles/CategoriesFilter.css";

class CategoriesFilter extends Component {
  constructor() {
    super();

    this.state = {
      stage: 0
    };

    this.handleAnimationEnd = this.handleAnimationEnd.bind(this);
    this.handleButtonClick = this.handleButtonClick.bind(this);
    this.handleOutsideClick = this.handleOutsideClick.bind(this);
  }

  getLifecycleStage() {
    const stages = {
      0: "inactive",
      1: "fade-in",
      2: "active",
      3: "fade-out"
    };

    return stages[this.state.stage];
  }

  handleButtonClick() {
    this.setState(({ stage }) => ({
      stage: ++stage
    }));
  }

  handleOutsideClick() {
    const stage = this.getLifecycleStage();

    if (stage !== "inactive") {
      this.setState({
        stage: 3
      });
    }
  }

  handleAnimationEnd(e) {
    const { animationName } = e;
    const stage = this.getLifecycleStage();

    if (animationName === "fadeIn" && stage === "fade-in") {
      this.setState(({ stage }) => ({
        stage: ++stage
      }));
    }

    if (animationName === "fadeOut" && stage === "fade-out") {
      this.setState({
        stage: 0
      });
    }
  }

  render() {
    const stage = this.getLifecycleStage();
    const buttonClasses = classNames({
      btn: true,
      "btn-round-big": true,
      active: stage === "fade-in" || stage === "active"
    });

    return (
      <OutsideClickHandler
        onOutsideClick={this.handleOutsideClick}
        disabled={stage === "inactive"}
      >
        <div className="CategoriesFilter">
          <button
            className={buttonClasses}
            onClick={this.handleButtonClick}
            title="Select categories"
          >
            <FilterList />
          </button>
          {stage !== "inactive" ? (
            <CategoriesList
              animationEndHandler={this.handleAnimationEnd}
              stage={stage}
            />
          ) : null}
        </div>
      </OutsideClickHandler>
    );
  }
}

export default CategoriesFilter;
