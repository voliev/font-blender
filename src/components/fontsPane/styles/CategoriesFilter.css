.CategoriesFilter {
  align-items: center;
  display: flex;
  grid-area: icon;
  height: 100%;
  position: relative;
  width: 100%;
}

.CategoriesFilter .btn {
  background-color: transparent;
  color: var(--blue);
}
