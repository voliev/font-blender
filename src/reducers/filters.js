import {
  SET_CATEGORIES_FILTER,
  SET_TITLE_FILTER,
  SET_SORTING_ORDER,
  SET_SORTING_PARAM
} from "../constants";

const initialState = {
  title: "",
  categories: {
    serif: true,
    "sans-serif": true,
    display: true,
    monospace: true,
    handwriting: true
  },
  sorting: {
    param: "popularity",
    order: "asc"
  }
};

export default (state = initialState, action) => {
  switch (action.type) {
    case SET_TITLE_FILTER:
      return {
        ...state,
        title: action.title
      };
    case SET_CATEGORIES_FILTER:
      return {
        ...state,
        categories: {
          ...state.categories,
          [action.category]: !state.categories[action.category]
        }
      };
    case SET_SORTING_PARAM:
      return {
        ...state,
        sorting: {
          ...state.sorting,
          param: action.param
        }
      };
    case SET_SORTING_ORDER:
      return {
        ...state,
        sorting: {
          ...state.sorting,
          order: action.order === "asc" ? "des" : "asc"
        }
      };
    default:
      return state;
  }
};

export const getFilters = state => state.filters;
