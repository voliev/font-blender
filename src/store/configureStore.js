import { createStore, combineReducers, applyMiddleware, compose } from "redux";
import thunk from "redux-thunk";
import fonts from "../reducers/fonts";
import filters from "../reducers/filters";
import selectedFonts from "../reducers/selectedFonts";
import showcase from "../reducers/showcase";
import styles from "../reducers/styles";
import notifications from "../reducers/notifications";

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export default () => {
  const store = createStore(
    combineReducers({
      fonts,
      filters,
      selectedFonts,
      showcase,
      styles,
      notifications
    }),
    composeEnhancers(applyMiddleware(thunk))
  );

  return store;
};
