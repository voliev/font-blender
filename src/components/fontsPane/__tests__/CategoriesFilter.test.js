import React from "react";
import { mount, shallow } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import OutsideClickHandler from "react-outside-click-handler";
import CategoriesFilter from "../CategoriesFilter";
import filters from "../../../fixtures/filters";

const setup = (renderFn = shallow, provideStore = false) => {
  if (provideStore) {
    const mockStore = configureStore();
    const store = mockStore({ filters });

    return {
      wrapper: renderFn(
        <Provider store={store}>
          <CategoriesFilter />
        </Provider>
      )
    };
  }

  return {
    handleButtonClick: jest.spyOn(
      CategoriesFilter.prototype,
      "handleButtonClick"
    ),
    handleOutsideClick: jest.spyOn(
      CategoriesFilter.prototype,
      "handleOutsideClick"
    ),
    handleAnimationEnd: jest.spyOn(
      CategoriesFilter.prototype,
      "handleAnimationEnd"
    ),
    wrapper: renderFn(<CategoriesFilter />)
  };
};

afterEach(() => {
  jest.clearAllMocks();
});

it("should render without crashing", () => {
  const { wrapper } = setup();

  expect(wrapper).toMatchSnapshot();
});

it("should set corresponding button classes", () => {
  const { wrapper } = setup();

  expect(wrapper.find("button").hasClass("btn")).toBeTruthy();
  expect(wrapper.find("button").hasClass("active")).toBeFalsy();

  wrapper.setState({ stage: 1 });

  expect(wrapper.find("button").hasClass("btn")).toBeTruthy();
  expect(wrapper.find("button").hasClass("active")).toBeTruthy();

  wrapper.setState({ stage: 2 });

  expect(wrapper.find("button").hasClass("btn")).toBeTruthy();
  expect(wrapper.find("button").hasClass("active")).toBeTruthy();

  wrapper.setState({ stage: 4 });

  expect(wrapper.find("button").hasClass("btn")).toBeTruthy();
  expect(wrapper.find("button").hasClass("active")).toBeFalsy();
});

it("should change stage on button click", () => {
  const { wrapper } = setup(shallow);

  expect(wrapper.state("stage")).toBe(0);

  wrapper.find("button").prop("onClick")();

  expect(wrapper.state("stage")).toBe(1);
});

it("should call handler on button click", () => {
  const { wrapper, handleButtonClick } = setup();

  wrapper.find("button").simulate("click");

  expect(handleButtonClick).toHaveBeenCalledTimes(1);
});

it("should handle click outside of the filter", () => {
  const { wrapper, handleOutsideClick } = setup();

  wrapper.setState({ stage: 1 });

  expect(wrapper.state("stage")).toBe(1);

  wrapper.find(OutsideClickHandler).prop("onOutsideClick")();

  expect(handleOutsideClick).toHaveBeenCalledTimes(1);
  expect(wrapper.state("stage")).toBe(3);
});

it("should not change the state when the filter is inactive", () => {
  const { wrapper, handleOutsideClick } = setup();

  wrapper.find(OutsideClickHandler).prop("onOutsideClick")();

  expect(handleOutsideClick).toHaveBeenCalledTimes(1);
  expect(wrapper.state("stage")).toBe(0);
});

it("should change state on button click", () => {
  const { wrapper } = setup(mount, true);

  expect(wrapper.find(CategoriesFilter).state().stage).toBe(0);

  wrapper.find("button").simulate("click");

  expect(wrapper.find(CategoriesFilter).state().stage).toBe(1);
});

it("should handle animation end", () => {
  const { wrapper } = setup();

  wrapper.setState({ stage: 1 });
  wrapper.instance().handleAnimationEnd({
    animationName: "fadeIn"
  });

  expect(wrapper.state("stage")).toBe(2);

  wrapper.setState({ stage: 3 });
  wrapper.instance().handleAnimationEnd({
    animationName: "fadeOut"
  });

  expect(wrapper.state("stage")).toBe(0);
});
