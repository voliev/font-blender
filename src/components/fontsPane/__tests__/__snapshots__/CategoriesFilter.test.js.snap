// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render without crashing 1`] = `
<OutsideClickHandler
  disabled={true}
  display="block"
  onOutsideClick={[Function]}
  useCapture={true}
>
  <div
    className="CategoriesFilter"
  >
    <button
      className="btn btn-round-big"
      onClick={[Function]}
      title="Select categories"
    >
      <FilterListIcon />
    </button>
  </div>
</OutsideClickHandler>
`;
